platform-sdk/swirlds-common/src/main/java/com/swirlds/common/merkle/proof/algorithms/StateProofTreeBuilder.java

```diff
@@ -0,0 +1,305 @@
+        if (payloads.isEmpty()) {
+            throw new IllegalArgumentException("payloads must not be empty");
+        }
+        for (final MerkleLeaf leaf : payloads) {
+            if (leaf == null) {
+                throw new IllegalArgumentException("payloads are not permitted to contain null leaves");
+            }
+        }
+        return payloads;
+    }
+
+    /**
+     * Do some basic sanity checks on a provided signature map and convert to a sorted list of {@link NodeSignature}s.
+     *
+     * @param unprocessedSignatures the signatures to process
+     * @return the processed signatures
+     */
+    @NonNull
+    public static List<NodeSignature> processSignatures(@NonNull final Map<NodeId, Signature> unprocessedSignatures) {
+        if (unprocessedSignatures.isEmpty()) {
+            throw new IllegalArgumentException("signatures must not be empty");
+        }
+
+        final List<NodeSignature> signatures = new ArrayList<>(unprocessedSignatures.size());
+        for (final Map.Entry<NodeId, Signature> entry : unprocessedSignatures.entrySet()) {
+            if (entry.getValue() == null) {
+                throw new IllegalArgumentException("signatures are not permitted to contain null values");
+            }
+            signatures.add(new NodeSignature(entry.getKey(), entry.getValue()));
+        }
+
+        Collections.sort(signatures);
+        return signatures;
+    }
+
+    /**
+     * Check if a merkle node is an ancestor of at least one of the payloads. Merkle nodes are considered to be
+     * ancestors to themselves.
+     *
+     * @param potentialAncestor the potential ancestor
+     * @param payloads          the payloads
+     * @return true if the potential ancestor is an ancestor of at least one payload, otherwise false
+     */
+    private static boolean isAncestorOfPayload(
+            @NonNull final MerkleNode potentialAncestor, @NonNull final List<MerkleLeaf> payloads) {
+
+        for (final MerkleLeaf payload : payloads) {
+            if (potentialAncestor.getRoute().isAncestorOf(payload.getRoute())) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Given a merkle root and a list of payloads, extract a list of merkle nodes that are ancestors of at least one of
+     * the payloads. Nodes are returned in reverse post-ordered depth first traversal order.
+     *
+     * @param merkleRoot the merkle root
+     * @param payloads   the payloads
+     * @return a list of merkle nodes that are ancestors of at least one of the payloads
+     * @throws IllegalStateException if the payloads and the merkle tree are inconsistent with each other
+     */
+    private static List<MerkleNode> getMerkleNodesForStateProofTree(
+            @NonNull final MerkleNode merkleRoot, @NonNull final List<MerkleLeaf> payloads) {
+
+        // Note: we use a reverse post-ordered depth first traversal here intentionally. This causes a node's children
```

Comment:
maybe move to JavaDoc?

Reply:
Removed comment since it is already covered in the javadoc (more or less).